rebasehelper/build_log_hooks/files_build_log_hook.py

```diff
@@ -0,0 +1,169 @@
+
+            elif next_line_is_missing:
+                if li.startswith('/'):
+                    files.add(line.strip())
+                else:
+                    next_line_is_missing = False
+
+            elif 'Installed (but unpackaged) file(s) found' in li:
+                # The following files are missing in the SPEC file.
+                next_line_is_missing = True
+                error_type = 'missing'
+
+        return error_type, files
+
+    @classmethod
+    def _get_best_matching_files_section(cls, rebase_spec_file, file):
+        """Finds a section with a file that has the closest match with
+        the given file.
+
+        Args:
+            rebase_spec_file (specfile.SpecFile): Rebased SpecFile object.
+            file (str): Path to the file to be classified.
+
+        Returns:
+            str: Name of the section containing the closest matching file.
+
+        """
+        best_match = ''
+        best_match_section = ''
+        for sec_name, sec_content in six.iteritems(rebase_spec_file.spec_content.sections):
+            if sec_name.startswith('%files'):
+                for line in sec_content:
+                    new_best_match = difflib.get_close_matches(file, [best_match, MacroHelper.expand(line)])
+                    if new_best_match:
+                        # the new match is a closer match
+                        if new_best_match[0] != best_match:
+                            best_match = new_best_match[0]
+                            best_match_section = sec_name
+
+        return best_match_section
+
+    @classmethod
+    def _correct_missing_files(cls, rebase_spec_file, files):
+        """Adds files found in sources to the SPEC file."""
+        macros = [m for m in rebase_spec_file.macros if m['name'] in MacroHelper.MACROS_WHITELIST]
+        macros = MacroHelper.expand_macros(macros)
+        # ensure maximal greediness
+        macros.sort(key=lambda k: len(k['value']), reverse=True)
+
+        for file in files:
+            section = cls._get_best_matching_files_section(rebase_spec_file, file)
+            if section:
+                substituted_path = MacroHelper.substitute_path_for_macro(file, macros)
+                rebase_spec_file.spec_content.sections[section].append(substituted_path)
+            else:
+                logger.info('Could not correctly add file %s to the SPEC file.', file)
+
+    @classmethod
+    def _correct_deleted_files(cls, rebase_spec_file, files):
+        """Removes files missing in sources from the SPEC file."""
+        for sec_name, sec_content in six.iteritems(rebase_spec_file.spec_content.sections):
+            if sec_name.startswith('%files'):
+                modified_section = []
+                for line in sec_content:
+                    file_is_ok = True
+                    if line != '':
+                        for file in files:
+                            if MacroHelper.expand(line).split(' ')[-1] in file:
```

Comment:
The expansion and splitting should happen before the loop, it's pointless to do it in every iteration.

Reply:
Also, why `in`?

Reply:
It expands and splits lines of the SPEC file so in each iteration it expands and splits one unique line from the spec. The files in build log are already expanded.

Also, the paths from build log contain build directory path, therefore the specfile path is a subset of the log path. I think it will be optimal to strip build directory from the path and then compare using fnmatch.

Reply:
> It expands and splits lines of the SPEC file so in each iteration it expands and splits one unique line from the spec. The files in build log are already expanded.

Are you saying that in each iteration of `for file in files:` loop `line` has different value?

> Also, the paths from build log contain build directory path, therefore the specfile path is a subset of the log path. I think it will be optimal to strip build directory from the path and then compare using fnmatch.

Yes, that sounds much better.

Reply:
Good point, I missed that one, thanks

Reply:
Two more things:

- there can be more paths on the same line, see [libtiff.spec](https://src.fedoraproject.org/rpms/libtiff/blob/b9f5e1a576c4a7e95050e3380c86a446fd4b891f/f/libtiff.spec#_169) for example
- the macro expansion will expand for example `%license COPYING` into e.g. `GPLv2+ COPYING`, which could be fine in theory, but it would be better to expand only the individual paths